### PR TITLE
nixos/paperless: set OMP_NUM_THREADS=1 by default

### LINKED ISSUE #240591: paperless-ngx: classifier training hangs & times out
### Describe the bug

Directly related to this discussion: https://github.com/paperless-ngx/paperless-ngx/discussions/2373

paperless-ngx runs a hourly `documents.tasks.train_classifier` celery beat task.  This is supposed to take a few minutes on most systems, but on NixOS (for some users, at least) the task runs forever and is eventually killed by the celery timeout.  The affected worker process doesn't respond to SIGTERM or even SIGQUIT; only SIGKILL can interrupt it.

### Steps To Reproduce
1. Set up paperless-ngx and add some documents to it
2. Wait a while(?) or manually trigger the train_classifier task
3. It times out after 30 minutes, probably.

### Expected behavior
The classifier training task should finish successfully in a few minutes or less.

### Additional context
See https://github.com/paperless-ngx/paperless-ngx/discussions/2373 for more background.  This appears to be specific to paperless-ngx on NixOS.

### Notify maintainers

@lukegb @gador @erikarvstedt @Flakebi 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.83, NixOS, 23.05 (Stoat), 23.05.git.9790f3242da2M`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(benley): `""`
 - channels(root): `"home-manager-23.05.tar.gz, nixos-23.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by Atemu**
Fixes https://github.com/NixOS/nixpkgs/issues/240591

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/misc/paperless.nix [hunk 1]
```diff
@@ -199,6 +201,20 @@ in
     };
 
     package = mkPackageOption pkgs "paperless-ngx" { };
+
+    opemMPThreadingWorkaround = mkEnableOption ''
```
Comment (leona-ya) [inline nixos/modules/services/misc/paperless.nix:205]:
```suggestion
    openMPThreadingWorkaround = mkEnableOption ''
```

**Review by leona-ya — COMMENTED** [review]
One small typo, otherwise looks good (until we find the solution)

**Comment by leona-ya** [comment]
We probably also want to backport this, right?

**Comment by Atemu** [comment]
Oh my, how embarrassing.

I'd like to wait a week or two with the backport just in case this breaks something super weird for someone out there but it really shouldn't.

File: nixos/modules/services/misc/paperless.nix [hunk 2]
```diff
@@ -199,6 +201,20 @@ in
...
+
+      Paperless uses OpenBLAS via scikit-learn for document classification.
+
+      The default is to use threading for OpenMP but this would cause the
+      document classifier to spin on one core seemingly indefinitely if there
+      are large amounts of classes per classification; causing it to
+      effectively never complete due to running into timeouts.
+
+      This sets `OMP_NUM_THREADS` to `1` in order to mitigate the issue. See
+      https://github.com/NixOS/nixpkgs/issues/240591 for more information.
```
Comment (ambroisie) [inline nixos/modules/services/misc/paperless.nix:216]: Small suggestion, can be merged without it:

```suggestion
      [#240591](https://github.com/NixOS/nixpkgs/issues/240591) for more information.
```

**Review by ambroisie — APPROVED** [review]

**Review by leona-ya — APPROVED** [review]

File: nixos/modules/services/misc/paperless.nix [hunk 3]
```diff
@@ -27,6 +27,8 @@ let
       name = "paperless_ngx_nltk_data";
       paths = pkg.nltkData;
     };
+  } // optionalAttrs (cfg.opemMPThreadingWorkaround) {
```
Comment (leona-ya) [inline nixos/modules/services/misc/paperless.nix:30]: ahhh you also need to change it here

**Review by leona-ya — CHANGES_REQUESTED** [review]

**Review by leona-ya — APPROVED** [review]

**Comment by Atemu** [comment]
https://github.com/NixOS/nixpkgs/pull/299289

**Pull request merged**